pyccel/parser/semantic.py

```diff
@@ -356,6 +368,34 @@ def check_for_variable(self, name):
         else:
             return self.scope.find(name, 'variables')
 
+    def check_for_attribute(self, name):
+        """
+        Search for an Attribute with the given name in the current scope.
+
+        Search for an Attribute with the given name in the current scope,
+        defined by the local and global Python scopes. Return None if not found.
+
+        Parameters
+        ----------
+        name : DottedName
+            The object describing the attribute.
+
+        Returns
+        -------
+        Variable
+            Returns the attribute if found or None.
+        """
+
+        if self.current_class:
+            attributes_list = list(self._current_class.attributes)
+            for key, value in enumerate(attributes_list):
+                if value.name == name.name[-1]:
+                    attributes_list[key] = value
+                    self._current_class.attributes = tuple(attributes_list)
+                    return value
+        return None
```

Comment:
I'm not sure if `self._current_class` is the most useful object here. As I understand it in the following example:
```python
class Line:
    def __init__(self, l):
        self.l = l

class Point:
    def func(self, l : Line):
         print(l.l)
```
in the function `func`, `self._current_class` is `Point`, but you want to use this function to find out about `Line`.

I suggest something like:
```suggestion
        assert isinstance(name, DottedName)
        prefix = self._visit(DottedName(*name.name[:-1]))
        class_def = prefix.cls_base
        attr_name = name.name[-1]
        try:
             value = next(a for a in class_def.attributes if a.name == attr_name)
        except StopIteration:
             value = None
        return value
```

Although if we are avoiding name collision problems, maybe it is simply:
```suggestion
        assert isinstance(name, DottedName)
        prefix = self._visit(DottedName(*name.name[:-1]))
        class_def = prefix.cls_base
        attr_name = name.name[-1]
        return class_def.scope.find(attr_name, 'variables')
```

Reply:
Where?

Reply:
Fixed now.

Reply:
This looks better. But the new code may raise the name collision problems that I mentioned above

Reply:
Fixed. [2fcd3f2](https://github.com/pyccel/pyccel/pull/1484/commits/2fcd3f251abe4f9584e5de7060ab6e20f05509c5)

Reply:
> Fixed. [2fcd3f2](https://github.com/pyccel/pyccel/pull/1484/commits/2fcd3f251abe4f9584e5de7060ab6e20f05509c5)

This does not fix the collisions it is just a different way of writing the same thing

Reply:
See #1503

Reply:
I found that this code : 
```python
class Line:
    def __init__(self, l):
        self.l = l

class Point:
    def __init__(self, l : Line):
        print(l.l)
```
I found that the problem comes from the `dtype` of the `lhs` of the variable `l.l` cause this `dtype` is saved as `line` but the classe name saved in the scope is `Line` so it returns an Error cause it s not the same string. I need just a hint if this change in the naming comes from the `semantic part` or the `syntactic part`.

Reply:
Huh. That's odd. It can't be the syntactic part because there are no types there.  It could be in ast.datatypes but I thought I'd got rid of the decapitalisation

Reply:
How are you printing the dtype? I think there is a difference between `str(dtype)` and `dtype.name`. I think you should be using `dtype.name` to look things up in the class not `dtype` which is a `CustomDataType` instance

Reply:
Fixed https://github.com/pyccel/pyccel/pull/1484/commits/da659e8c20e668003a7ee8480d3724d51b9179b0.